### DISCUSSION TITLE
Rebase with advanced options does change rebased commits committer

### DISCUSSION THREAD

**Discussion by andry81** [gitextensions/gitextensions#10007] [Q&A] [answered]
### Environment

- Git Extensions 3.5.4.12724
- Build 65f01f39982ccb121c4df7d4b00b506939553958
- Git 2.36.1.windows.1
- Microsoft Windows NT 6.1.7601 Service Pack 1
- .NET Framework 4.8.4069.0
- DPI 96dpi (no scaling)

### Issue description

If try to use the commit context menu: `Rebase current branch on -> Selected commit with advanced options...`

And use `Specific range` to remove a commit in between of a HEAD and a BASE commits. Then the resulted rebased commits will change the committer to one which is initiated the rebase.

I think it should not happen because in the git a commit remove is a rebase/rewrite operation without any other change. But instead the committer does change.

### Steps to reproduce

* Select commit as a BASE
* Use `Rebase current branch on -> Selected commit with advanced options...` context menu on the commit
* Select `Specific range` checkbox and set range of commits between a BASE commit and a HEAD commit (for example, to remove all the commits not in the range but after a BASE commit).
* Merge all commits while rebasing
* Observe changed committer

I think something like

`[x] Retain original committer` and `[x] Committer is author`

option with *selected by default* must be

The same for all other rebase options by default too.

**Comment by vbjay** [comment, marked as answer]
This is correct. You have created a new commit as you.  The author doesn't
change but yes, you are now the one who committed.

**Comment by gerhardol** [comment]
https://git-scm.com/docs/git-rebase

There is no possibility to retain committer for git-rebase.
If it was, it would be cheating, the committer is what you configure (so technically you could set the committer name/email before each individual commit is cherry-picked. Even if that could be implemented, I would vote no to that, it is not how Git is intended to be used. signed commits is not used so much because the committer is trusted).

**Comment by andry81** [comment]
Does it mean I can not delete a commit without change anything else?

**Comment by gerhardol** [comment]
> Does it mean I can not delete a commit without change anything else?

As the parents are a part of the sha for a commit, all rewrites will change the sha.
And committer will be set to the current user.

**Comment by andry81** [comment]
> And committer will be set to the current user.

I didn't want to change committers, I just wanted to delete a commit.

**Comment by pmiossec** [comment]
> I didn't want to change committers, I just wanted to delete a commit.

But that's how git works so I don't think you could go against it.
To sum up, you have 2 possibilities:

1. You don't want to or can't change the git history. You will have to do a revert of the commit you want to undo.
2. You want to change the history to remove the commit. In this case, you must admit that for all the commits based on the this commit, new commits, very similar, will be created (during the rebase because that's how it works) but with the committer information changed to contains the name of the user that rewrote the history (because, indeed, that's this user that created these new commits) . I think we could consider that normal for tracking purpose, no? 

With these information and constraints, choose the one you really want to apply. What you want is the advantages of the 2 solutions but that's not how git works...

**Comment by vbjay** [comment]
Think of it like this. Say you have 10 marbles on a string. Each marble has qualities that control how the next marble is cast and placed on string. If you alter( technically replace ) marble 5 then marbles 6-10 get recreated too.  You change anything on a commit and the sha hash changes. The hash of the parent is part of a commit. So 5 is replaced and it gets new sha. So 6 is recreated using new sha from 5 and so on. 


Read up on the difference between commiter and author.

**Reply by andry81** [reply]
> Think of it like this. Say you have 10 marbles on a string. Each marble has qualities that control how the next marble is cast and placed on string. If you alter( technically replace ) marble 5 then marbles 6-10 get recreated too. You change anything on a commit and the sha hash changes. The hash of the parent is part of a commit. So 5 is replaced and it gets new sha. So 6 is recreated using new sha from 5 and so on.

I don't bother about hashes. Its a system overhead maintained by the system. The committer is a user overhead. And I change it unexpectedly, because git designed so. That is not what I actually wanted to change. In truth, this is implicit change hidden from a user. And you will notice it only after make a rebase with default parameters. It should not behave like that in first place. That's the problem.

> Read up on the difference between commiter and author.

Author does not change. But the committer does change. I didn't commit the rebased commits in first place. It already has been done before me. And the GitHub will show the committer which actually is not me.

There is must be something like a branch-edit committer.

**Reply by pmiossec** [reply]
> should not behave like that in first place. That's the problem.

@andry81 we are trying to explain you why it behaves like that and that you cannot achieve what you expect.

If you really want to argue go give your arguments in the git mailing list telling them that what they did is wrong and why. I'm not sure they will be as patient as us.

The rule is simple : when you rebase, you recreate some commits so YOU are the new committer of these commits. Period. The author information is there to retain the information of the initial creator of the commit.
There is nothing much we can do. Especially GitExtensions that is just a wrapper around git commands.

Looks like a [Dunning-Kruger effect](https://en.m.wikipedia.org/wiki/Dunning%E2%80%93Kruger_effect) case...

**Reply by vbjay** [reply]
I downloaded python scripts from https://github.com/alkampfergit/GitGraphAutomation/tree/main/src into a folder.

In powershell I ran

```powershell
$PSDefaultParameterValues = @{'*:Encoding' = 'utf8'}
git init .
git config user.name "John Smith"
git config user.email 123@123.com

npm install playwright
echo "node_modules/
">.gitignore
echo "*.txt text
">.gitattributes
git add .
git commit -m "Initial Repo Setup"

echo "Initial work">work.txt
git add .
git commit -m "initial work"

echo "more work">>work.txt
git add .
git commit "more work"

echo "more work 2">>work.txt
git add .
git commit -m "more work 2"

echo "more work 3">>workother.txt
git add .
git commit -m "more work 3"

echo "more work 4">>work.txt
git add .
git commit -m "more work 4"

echo "more work 5">>work.txt
git add .
git commit -m "more work 5"


git tag -a BeforeRebase -m "Before rebase is done"
git branch Before
python .\gitRender.py --repo .\ --outhtml .\BeforeRebase.html --renderpng .\BeforeRebase.png

# represents another user doing the rebase
git config user.name "Sally Smith"
git config user.email 124@124.com

git rebase --onto HEAD~2^ HEAD~2
git branch After
git tag -a AfterRebase -m "After rebase is done"
python .\gitRender.py --repo .\ --outhtml .\AfterRebase.html --renderpng .\AfterRebase.png


```
![image](https://user-images.githubusercontent.com/3629489/169819849-f97a8b4d-fa7d-4c91-99b3-a73b76fc0270.png)

- Before rebase 
![image](https://user-images.githubusercontent.com/3629489/169820553-7c86388d-b60d-4bc3-b89c-69fa623583e8.png)
![image](https://user-images.githubusercontent.com/3629489/169820109-c8052eb7-42e4-4d57-bef8-8edd28dae6f2.png)
![image](https://user-images.githubusercontent.com/3629489/169819939-5fd7a219-9317-4a84-982e-3839e3472ac4.png)
- After Rebase 
![image](https://user-images.githubusercontent.com/3629489/169820626-057b2aad-65ca-457e-90e2-66fade595e27.png)
![image](https://user-images.githubusercontent.com/3629489/169820335-d0e48896-3c67-4a9e-b953-7da50655c050.png)
![image](https://user-images.githubusercontent.com/3629489/169820005-8c79ad5f-c5d4-43cb-8cf1-95ee970d8cff.png)

Do you now get it?  This example is why I use git signing.  I have the following in my global git config.

```
[user]
	name = Jay Asbury
	email = vbjay.net@gmail.com
	signingkey = 223DF2DDA3ECCBBF
[commit]
	gpgsign = true
[gpg]
	program = C:/Program Files (x86)/GnuPG/bin/gpg.exe

```

![image](https://user-images.githubusercontent.com/3629489/169823768-08148f1a-53b3-4cf3-9d49-581e60b371de.png)

**Reply by vbjay** [reply]
Updated above with more info.  Might want to make sure you view this in github and not just email.

**Reply by vbjay** [reply]
On the note of signing.  Signing is better used if all sign.  Then in your CI/CD, you can verify commits are signed and halt build if signature isn't there or is invalid.  Sure would have helped with Solarwinds attack.

**Reply by andry81** [reply]
>@andry81 we are trying to explain you why it behaves like that and that you cannot achieve what you expect.
>
>Looks like a [Dunning-Kruger effect](https://en.m.wikipedia.org/wiki/Dunning%E2%80%93Kruger_effect) case...

Oh, very thank of you. Without your advice I would never get to knowledge for myself.. already.

>If you really want to argue go give your arguments in the git mailing list telling them that what they did is wrong and why. I'm not sure they >will be as patient as us.

I have to give my graceful arguments wherever I want. This is internet and this is how a conversation does move along in it, deal with that. And if I want I will bring the arguments upon the git folks. Why? Because I CAN. And don't glue "us" to yourself, that's not gonna give away.

**Reply by vbjay** [reply]
Oh can't WAIT to read their response.  LOL.  Please share your request to them and their response. 
![popcorn](https://user-images.githubusercontent.com/3629489/169850540-c3761fa8-4e4d-4ddd-9591-dc257062c7e3.gif)

**Reply by vbjay** [reply]
https://git.wiki.kernel.org/index.php/GitCommunity  In case you need it.

**Reply by vbjay** [reply]
I very clearly demonstrated what a rebase does visually for you.  The point is that the history was rewritten.  This means NEW commits were created.  this means .... the person doing the rebase created the NEW commits.  How you don't get this....is beyond me.

**Reply by vbjay** [reply]
It would behoove you to read this.  Pay attention to parent/child relationship.  

https://blog.thoughtram.io/git/2014/11/18/the-anatomy-of-a-git-commit.html

**Reply by andry81** [reply]
> I very clearly demonstrated what a rebase does visually for you. The point is that the history was rewritten. This means NEW commits were created. this means .... the person doing the rebase created the NEW commits. How you don't get this....is beyond me.

Initial operation was delete followed by rebase. It's quite not a rewrite or a new.

**Reply by vbjay** [reply]
A rebase IS a rewrite.

**Reply by vbjay** [reply]
The delete was a rebase.   Follow along.  


A commit has meta data like author, commit date root tree hash...and others.  One of the pieces that makes up a commit is the collection of parent commit hashes( think merge commits where you have two or more commits being parents.)  All of this meta data is hashed together to get the id of the commit or sha.  This is how a branch works.  You just have a pointer to a commit and it points to its parent and so on and so on till you get to initial commit which doesn't have a parent.  A branch is just a name that points to a sha. 

![image](https://user-images.githubusercontent.com/3629489/169858352-c2f76e55-65bd-4aca-b48b-cba2e0a98880.png)
![image](https://user-images.githubusercontent.com/3629489/169858440-86802151-0dc3-47e1-be7f-7c46645e33ee.png)

So the ref points to the head commit of the ref and that commit points to one or more parents and each of those point to their parents.  Wash. Rinse. Repeat.  

So if you alter the commit by taking an existing commit by running a rebase you end up changing overwriting the  history because you changed the parent.   more work 3 commit f8dc56c6e0f23199e1a19ae503fff1989b427768 is pointed to by more work 4 commit in before branch.  May show short hash but it still is pointing to full sha.
![image](https://user-images.githubusercontent.com/3629489/169859185-59cdd1eb-8b68-4fb2-8e31-b70662b224c5.png)

So since the drop of f8dc56c6e0f23199e1a19ae503fff1989b427768 happened then the new commit has a parent of e218593250691aae6126f5f0961db1b98fc37149  
![image](https://user-images.githubusercontent.com/3629489/169859722-f3886e0a-021e-459a-b874-e92eab7a705a.png)

Since you should by now get the idea that the more work 4 commit was rewritten since the parent and technically the commit time has changed...it created  a new commit.  But hmm more work 5 c8a669b1d0728a1f96efe94686735a5dd0c6d164 commit was pointing to 45f390e60eea2fae22b13eddd78eeeb39199e31d.  So that requires c8a669b1d0728a1f96efe94686735a5dd0c6d164 to be rewritten which it was to be  56994f1c8890a31645b7ba486195b6946acc7090.


Do you finally get it?

**Reply by vbjay** [reply]
By the way what command ran to do the delete?  

# **REBASE**

**Reply by vbjay** [reply]
Technically what a rebase does. 
https://git-scm.com/docs/git-rebase

> The current branch is reset to <upstream>, or <newbase> if the --onto option was supplied. This has the exact same effect as git reset --hard <upstream> (or <newbase>). ORIG_HEAD is set to point at the tip of the branch before the reset.

![image](https://user-images.githubusercontent.com/3629489/169864206-4e7e7ea8-4dff-46ac-ad26-000247c61126.png)

Detailed what --onto does 
https://stackoverflow.com/a/29916361/815813

**Reply by vbjay** [reply]
Highly recommend going through these to see how git works and what it does as you run different commands

https://learngitbranching.js.org/

**Reply by andry81** [reply]
To have one good VCS system actually you don't need a block chain. Simple hash per commit is enough to check the difference with another copy of commits tree. Even more, you don't need hashing at all, because comparison exist as a standalone, no need to make a hash to compare anything. No parent hash, not child hash, not even a hash. A hash only speedups a comparison, but not creates it.

So all that said before is just that: we have hashes to hash hashes to hold hashes to have hashes and loop it again.

**Reply by vbjay** [reply]
Looking forward to seeing your brand new vcs.
